### ISSUE TITLE
matrix factorization

### ISSUE THREAD

**Issue by skkulsc** [skkulsc/skku#8]
## Model 설명

- User와 News에 대한 table이 있을 때, 존재하는 data에 대해서 Matrix Factorization을 수행한 뒤, user가 읽지 않은 news에 대한 점수를 예측.

- table에는 평점 대신, user가 특정 news를 읽은 횟수가 기록되어 있음. 평점과 마찬가지로 user가 여러 번 읽은 news 일수록 그 주제에 흥미가 있으리라 예측할 수 있음.

- matrix factorization은 Tensorflow framework를 사용함.

- 해당 모델은 아래의 식이 최소가 되는 user feature matrix, item feature matrix, user bias, item bias를 찾는 것이 목표임

![mf](https://user-images.githubusercontent.com/37528988/47255326-562b3080-d4aa-11e8-9594-e9380b3967f7.png)

- Database에 어떤 user가 어떤 news를 읽었는지에 대한 정보가 저장되어 있음. Python의 sqlAlchemy 라이브러리를 이용해 local의 MySQL에 접속하여 DB에서 해당 정보를 불러옴. 불러온 정보를 다시 Numpy 형태로 변환하여 Matrix Factorization을 수행함.

- MovieLens (https://grouplens.org/datasets/movielens/) 의 dataset을 이용해 model의 성능을 측정함

-  전체 dataset에서 sklearn 라이브러리의 StratifiedKFold 모듈을 사용해 10개의 fold로 나눔(단순한 KFold 모듈을 사용하는 경우, 어느 user의 data는 training data에 쏠려있고, 어느 user의 data는 testing data에 쏠려있는 경우가 발생함). 10개의 fold 중에서 마지막 1개의 fold를 testing set으로 사용하고, 나머지 9개의 fold를 다시 training set과 validation set으로 나눠서(8 : 1) model을 학습함.

![cross validation](https://user-images.githubusercontent.com/37528988/47278668-a4087b80-d605-11e8-8b26-0adc60560168.png)

- validation RMSE의 추이에 따라 초기에 설정된 learning rate에 0.96을 곱해줘 learning rate를 줄여나감.

- 위의 식과 같이, 기본 loss에 regularization 식을 추가해 최종 loss를 만듬.

## Model 동작

- 아래 그림과 같이, matrix factorization으로 인해 나눠지는 두 개의 matrix W, H를 선언함(W는 user에 대한 matrix, H는 item에 대한 matrix) .

http://hameddaily.blogspot.com/2016/12/simple-matrix-factorization-with.html

![matrix_factorization](https://user-images.githubusercontent.com/37528988/47278831-d49ce500-d606-11e8-8f0f-5d69c9c3f7c8.png)

- W와 H에 각각 bias를 concat하여 최종적으로 W의 shape는 [num_users, rank + 1], H의 shape는 [num_items, rank + 1]이 됨.

----------------------------------------------------------------------------------------------------------

    W = tf.Variable(tf.truncated_normal([num_users, self.rank], stddev = 0.1) # user 
    W_bias = tf.Variable(tf.truncated_normal([num_users, 1], stddev = 0.1)) # user bias
            
    H = tf.Variable(tf.truncated_normal([num_items, self.rank], stddev = 0.1) # item
    H_bias = tf.Variable(tf.truncated_normal([num_items, 1], stddev = 0.1) # item bias
        
    user_bias = tf.concat([W, W_bias], axis = 1)
    item_bias = tf.concat([H, H_bias], axis = 1)

----------------------------------------------------------------------------------------------------------

- W와 H.T를 곱해 [num_users, num_items] shape를 가진 Matrix를 구한 다음, 실제 rating과 비교함

----------------------------------------------------------------------------------------------------------

    pred_pref = tf.matmul(user_bias, item_bias, transpose_b = True) # shape = [num_users, num_items]

----------------------------------------------------------------------------------------------------------
       
- 실제 rating과 예측한 rating과의 오차와 regularization term을 이용해 최종 loss를 구함

----------------------------------------------------------------------------------------------------------

    sub = tf.subtract(rates_list, get_results + global_mean_tensor)  
   
    basic_loss = tf.sqrt(tf.reduce_sum(tf.pow(sub, 2)))   

    lda = tf.constant(self.reg_value)
    norm_sums = tf.nn.l2_loss(W) + tf.nn.l2_loss(W_bias) + tf.nn.l2_loss(H) + tf.nn.l2_loss(H_bias)
    regul = tf.multiply(norm_sums, lda, name = 'regularizer')     
     
    final_loss = basic_loss + regul

----------------------------------------------------------------------------------------------------------

- model의 성능은 RMSE를 사용함

https://www.includehelp.com/ml-ai/root-mean-square%20error-rmse.aspx

![rmse](https://user-images.githubusercontent.com/37528988/47278993-3a3da100-d608-11e8-82c1-db9890d4936e.jpg)

----------------------------------------------------------------------------------------------------------

    rmse = tf.sqrt(tf.divide(tf.reduce_sum(tf.pow(sub, 2)), num_rates))

----------------------------------------------------------------------------------------------------------

## MovieLens (size 100K) data를 이용해 성능 측정(30 Epoch)

- training (8 folds for training, 1 folds for validation)

----------------------------------------------------------------------------------------------------------

Epoch = [   0] Training RMSE = [   1.026829] Val RMSE = [   1.036792] Best Val RMSE = [   1.036792]
Epoch = [ 10] Training RMSE = [   0.558438] Val RMSE = [   0.588797] Best Val RMSE = [   0.588797]
Epoch = [ 20] Training RMSE = [   0.094157] Val RMSE = [   0.107643] Best Val RMSE = [   0.107643]
Epoch = [ 29] Training RMSE = [   0.002792] Val RMSE = [   0.002414] Best Val RMSE = [   0.002414]

----------------------------------------------------------------------------------------------------------

- TestSet RMSE :  [0.0027418043]
